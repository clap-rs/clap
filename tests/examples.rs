use insta;
use std::ffi::OsStr;
use std::fs;
use std::process::{Command, Output};

fn run_example<S: AsRef<str>>(name: S, args: &[&str]) -> Output {
    let mut all_args = vec![
        "run",
        "--example",
        name.as_ref(),
        "--features",
        "yaml unstable",
        "--",
    ];
    all_args.extend_from_slice(args);

    Command::new(env!("CARGO"))
        .args(all_args)
        .output()
        .expect("failed to run example")
}

#[test]
fn examples_match_snapshots() {
    let example_paths = fs::read_dir("examples")
        .expect("couldn't read examples directory")
        .map(|result| result.expect("couldn't get directory entry").path())
        .filter(|path| path.is_file() && path.extension().and_then(OsStr::to_str) == Some("rs"));

    let mut example_count = 0;
    for path in example_paths {
        example_count += 1;

        let example_name = path
            .file_stem()
            .and_then(OsStr::to_str)
            .expect("unable to determine example name");

        let help = {
            let help_output = run_example(example_name, &["--help"]);
            assert!(
                help_output.status.success(),
                "{} example failed",
                example_name
            );
            String::from_utf8(help_output.stdout).expect("help is not valid utf-8")
        };

        insta::assert_snapshot!(format!("{} --help", example_name), help);
    }
    assert!(example_count > 0);
}
