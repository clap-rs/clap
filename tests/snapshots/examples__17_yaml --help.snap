---
source: tests/examples.rs
expression: help
---
yaml_app 1.0
Kevin K. <kbknapp@gmail.com>
an example using a .yaml file to build a CLI

USAGE:
    17_yaml [OPTIONS] [pos] [SUBCOMMAND]

ARGS:
    <pos>    example positional argument from yaml [possible values: fast, slow]

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -F <flag>...                   demo flag argument
        --max-vals <maxvals>...    you can only supply a max of 3 values for me!
        --min-vals <minvals>...    you must supply at least two values to satisfy me
        --mode <mode>              shows an option with specific values [possible values: vi, emacs]
        --mult-vals <one> <two>    demos an option which has two named values
    -o, --option <opt>...          example option argument from yaml

SUBCOMMANDS:
    help      Prints this message or the help of the given subcommand(s)
    subcmd    demos subcommands from yaml

